File: src/203. Remove Linked List Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        
        if (head == nullptr)
            return head;
        
        ListNode* prev = nullptr;
        ListNode* now = head;
        ListNode* next = now->next;
        
        while (now != nullptr) {
            if (now->val == val) {
                if (prev != nullptr) {
                    prev->next = now->next;
                }
                else {
                    head = now->next;
                }
                delete(now);
                now = nullptr;
            }
            
            prev = (now == nullptr) ? prev : now;
            now = next;            
            if (next != nullptr)
                next = next->next;
        }
        
        return head;
    }
};

File: src/203. Remove Linked List Elements.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode $head
     * @param Integer $val
     * @return ListNode
     */
    function removeElements($head, $val) {
        $prev = &$head;
        $now = &$head;

        while (true ){
            if ($now->val == $val) {
                if ($now == $head) {
                    $head = $now->next;
                }
                else {
                    $next = $now->next;
                    $prev->next = $next;
                }
            }

            if ($now->next == null) {
                break;
            }
            else {
                $now = $now->next;
                $prev = $now;
            }
        }

        return $head;
    }
}